core/src/main/java/io/questdb/cutlass/http/HttpSessionStore.java

```diff
@@ -0,0 +1,171 @@
+     */
+    @TestOnly
+    @Nullable
+    ObjList<SessionInfo> getSessions(@NotNull CharSequence principal);
+
+    /**
+     * Overrides the token generator used to generate session ids.
+     * Useful for testing.
+     */
+    @TestOnly
+    default void setTokenGenerator(TokenGenerator tokenGenerator) {
+    }
+
+    /**
+     * Lookup the session associated with the session id.
+     * If the session is valid, it will be returned.
+     * If no valid session exists for the session id, or the session already expired, there is no session returned.
+     *
+     * @param sessionId   session id to verify
+     * @param httpContext HTTP context associated with the user's connection
+     * @return session associated with the session id, or null if the session does not exist or expired
+     */
+    SessionInfo verifySessionId(@NotNull CharSequence sessionId, @NotNull HttpConnectionContext httpContext);
+
+    class SessionInfo implements PrincipalContext {
+        private final byte authType;
+        private final ObjList<CharSequence> groupsA = new ObjList<>();
+        private final ObjList<CharSequence> groupsB = new ObjList<>();
+        private final AtomicBoolean lock = new AtomicBoolean();
+        private final String principal;
+        private volatile long expiresAt;
+        private volatile ObjList<CharSequence> groups = groupsB;
+        private volatile long rotateAt;
+        private volatile String sessionId;
+
+        public SessionInfo(@NotNull String sessionId, String principal, @Nullable ObjList<CharSequence> groups, byte authType, long expiresAt, long rotateAt) {
+            this.sessionId = sessionId;
+            this.principal = principal;
+            this.authType = authType;
+            this.expiresAt = expiresAt;
+            this.rotateAt = rotateAt;
+
+            setGroups(groups);
+        }
+
+        @Override
+        public byte getAuthType() {
+            return authType;
+        }
+
+        public long getExpiresAt() {
+            return expiresAt;
+        }
+
+        @Override
+        public ObjList<CharSequence> getGroups() {
+            return groups;
+        }
+
+        @Override
+        public String getPrincipal() {
+            return principal;
+        }
+
+        public long getRotateAt() {
+            return rotateAt;
+        }
+
+        public String getSessionId() {
+            return sessionId;
+        }
+
+        public void setExpiresAt(long expiresAt) {
+            this.expiresAt = expiresAt;
+        }
+
+        public synchronized void setGroups(@Nullable ObjList<CharSequence> source) {
+            // ideally these would be compared as sets, but it is ok
+            // unlikely that the order of groups changing constantly
+            if (groups.equals(source)) {
+                return;
+            }
+
+            // select non-active list as target
+            final ObjList<CharSequence> target = groups == groupsA ? groupsB : groupsA;
```

Comment:
With frequent updates, this can still lead to race conditions. If someone holds ref to group A while 2 updates to group happen, it can read the list while this code populates it.

Usage of the same SessionInfo from multple threads is unlikely, I'd make `getGroups()` synchronized instead of this A, B swaps.